In [22]:
import pandas as pd
import numpy as np
from matplotlib import font_manager, rc
import matplotlib.pyplot as plt
import platform

if platform.system() == 'Windows':
# 윈도우인 경우
    font_name = font_manager.FontProperties(fname="c:/Windows/Fonts/malgun.ttf").get_name()
    rc('font', family=font_name)
else:    
# Mac 인 경우
    rc('font', family='AppleGothic')

In [23]:
def load_dataset(): 
    sales = pd.read_excel("sales.xlsx",header =1)
    view_rate = pd.read_excel("view_rate.xlsx",header =1, index_col= 0)
    return sales, view_rate

sales = load_dataset()[0]
view_rate = load_dataset()[1]

In [24]:
sales.head()

방송일시  노출(분)    마더코드    상품코드            상품명 상품군   판매단가  \
0 2019-01-01 06:00:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   
1 2019-01-01 06:00:00    NaN  100346  201079  테이트 여성 셀린니트3종  의류  39900   
2 2019-01-01 06:20:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   
3 2019-01-01 06:20:00    NaN  100346  201079  테이트 여성 셀린니트3종  의류  39900   
4 2019-01-01 06:40:00   20.0  100346  201072  테이트 남성 셀린니트3종  의류  39900   

         취급액  
0  2099000.0  
1  4371000.0  
2  3262000.0  
3  6955000.0  
4  6672000.0

In [25]:
view_rate.head()

2019-01-01  2019-01-02  2019-01-03  2019-01-04  2019-01-05  2019-01-06  \
시간대                                                                             
02:00         0.0         0.0       0.000         0.0         0.0         0.0   
02:01         0.0         0.0       0.000         0.0         0.0         0.0   
02:02         0.0         0.0       0.000         0.0         0.0         0.0   
02:03         0.0         0.0       0.014         0.0         0.0         0.0   
02:04         0.0         0.0       0.000         0.0         0.0         0.0   

       2019-01-07  2019-01-08  2019-01-09  2019-01-10  ...  2019-12-23  \
시간대                                                    ...               
02:00         0.0         0.0         0.0       0.000  ...         0.0   
02:01         0.0         0.0         0.0       0.005  ...         0.0   
02:02         0.0         0.0         0.0       0.005  ...         0.0   
02:03         0.0         0.0         0.0       0.005  ...         0.0   
02:04         0.0         0.0         0.0       0.005  ...         0.0   

       2019-12-24  2019-12-25  2019-12-26  2019-12-27  2019-12-28  2019-12-29  \
시간대                                                                             
02:00         0.0       0.000         0.0         0.0         0.0         0.0   
02:01         0.0       0.000         0.0         0.0         0.0         0.0   
02:02         0.0       0.000         0.0         0.0         0.0         0.0   
02:03         0.0       0.017         0.0         0.0         0.0         0.0   
02:04         0.0       0.017         0.0         0.0         0.0         0.0   

       2019-12-30  2019-12-31  2019-01-01 to 2019-12-31  
시간대                                                      
02:00         0.0       0.000                     0.003  
02:01         0.0       0.012                     0.003  
02:02         0.0       0.000                     0.004  
02:03         0.0       0.000                     0.004  
02:04         0.0       0.000                     0.004  

[5 rows x 366 columns]

In [26]:
# 무형 지우기
sales.drop(sales[sales["상품군"]=="무형"].index, inplace=True)

In [27]:
# 결측된 노출분 채우기
for i,row in sales.iterrows():
    val = row["노출(분)"]
    if i == 0 :
        continue
        
    if  np.isnan(val):
        if sales.loc[i,"방송일시"] == sales.loc[i-1,"방송일시"]:
            sales.loc[i,"노출(분)"] = sales.loc[i-1,"노출(분)"]
        else : 
            continue
            
sales.isnull().sum()

방송일시     0
노출(분)    0
마더코드     0
상품코드     0
상품명      0
상품군      0
판매단가     0
취급액      0
dtype: int64

In [48]:
# 판매단가 > 취급액 이면 취급액 = 0 

def custom(price, total):
    if price > total:
        return 0
    
    else:
        return total

sales["취급액"] = sales.apply(lambda x : custom(x["판매단가"], x["취급액"]), axis=1)

# 확인
print(sum((sales["판매단가"] > sales["취급액"]) & (sales['취급액'] != 0)))
sales

0


방송일시  노출(분)    마더코드    상품코드             상품명 상품군    판매단가  \
0     2019-01-01 06:00:00   20.0  100346  201072   테이트 남성 셀린니트3종  의류   39900   
1     2019-01-01 06:00:00   20.0  100346  201079   테이트 여성 셀린니트3종  의류   39900   
2     2019-01-01 06:20:00   20.0  100346  201072   테이트 남성 셀린니트3종  의류   39900   
3     2019-01-01 06:20:00   20.0  100346  201079   테이트 여성 셀린니트3종  의류   39900   
4     2019-01-01 06:40:00   20.0  100346  201072   테이트 남성 셀린니트3종  의류   39900   
...                   ...    ...     ...     ...             ...  ..     ...   
38299 2019-12-31 23:40:00   20.0  100448  201391   일시불쿠첸압력밥솥 6인용  주방  148000   
38300 2020-01-01 00:00:00   20.0  100448  201383  무이자쿠첸압력밥솥 10인용  주방  178000   
38301 2020-01-01 00:00:00   20.0  100448  201390  일시불쿠첸압력밥솥 10인용  주방  168000   
38302 2020-01-01 00:00:00   20.0  100448  201384   무이자쿠첸압력밥솥 6인용  주방  158000   
38303 2020-01-01 00:00:00   20.0  100448  201391   일시불쿠첸압력밥솥 6인용  주방  148000   

               취급액  
0        2099000.0  
1        4371000.0  
2        3262000.0  
3        6955000.0  
4        6672000.0  
...            ...  
38299   10157000.0  
38300   50929000.0  
38301  104392000.0  
38302   13765000.0  
38303   46608000.0  

[37372 rows x 8 columns]

In [54]:
# month, day, hour, minute, weekday 열 추가
sales['month'] = pd.DatetimeIndex(sales['방송일시']).month
sales['day'] = pd.DatetimeIndex(sales['방송일시']).day
sales['hour'] = pd.DatetimeIndex(sales['방송일시']).hour
sales['minute'] = pd.DatetimeIndex(sales['방송일시']).minute
sales['weekday'] = pd.DatetimeIndex(sales['방송일시']).weekday

In [55]:
sales

방송일시  노출(분)    마더코드    상품코드             상품명 상품군    판매단가  \
0     2019-01-01 06:00:00   20.0  100346  201072   테이트 남성 셀린니트3종  의류   39900   
1     2019-01-01 06:00:00   20.0  100346  201079   테이트 여성 셀린니트3종  의류   39900   
2     2019-01-01 06:20:00   20.0  100346  201072   테이트 남성 셀린니트3종  의류   39900   
3     2019-01-01 06:20:00   20.0  100346  201079   테이트 여성 셀린니트3종  의류   39900   
4     2019-01-01 06:40:00   20.0  100346  201072   테이트 남성 셀린니트3종  의류   39900   
...                   ...    ...     ...     ...             ...  ..     ...   
38299 2019-12-31 23:40:00   20.0  100448  201391   일시불쿠첸압력밥솥 6인용  주방  148000   
38300 2020-01-01 00:00:00   20.0  100448  201383  무이자쿠첸압력밥솥 10인용  주방  178000   
38301 2020-01-01 00:00:00   20.0  100448  201390  일시불쿠첸압력밥솥 10인용  주방  168000   
38302 2020-01-01 00:00:00   20.0  100448  201384   무이자쿠첸압력밥솥 6인용  주방  158000   
38303 2020-01-01 00:00:00   20.0  100448  201391   일시불쿠첸압력밥솥 6인용  주방  148000   

               취급액  month  day  hour  minute  weekday  
0        2099000.0      1    1     6       0        1  
1        4371000.0      1    1     6       0        1  
2        3262000.0      1    1     6      20        1  
3        6955000.0      1    1     6      20        1  
4        6672000.0      1    1     6      40        1  
...            ...    ...  ...   ...     ...      ...  
38299   10157000.0     12   31    23      40        1  
38300   50929000.0      1    1     0       0        2  
38301  104392000.0      1    1     0       0        2  
38302   13765000.0      1    1     0       0        2  
38303   46608000.0      1    1     0       0        2  

[37372 rows x 13 columns]

In [29]:
# 마지막 열의 값이 mena일까? 
test_lst= np.round(np.mean(view_rate.iloc[:,:-1].values,axis=1),3)==view_rate.iloc[:,-1].values
print(test_lst.all())
for i in np.argwhere(test_lst ==False):
    print("{}번째 열들의 평균은 ".format(i))
    print("마지막 열 직전까지의 값 {}".format(np.mean(view_rate.iloc[i,:-1].values)))
    print("실제값 : {}\n".format(view_rate.iloc[i,-1].values))

False
[93]번째 열들의 평균은 
마지막 열 직전까지의 값 0.002495890410958904
실제값 : [0.003]

[120]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0015013698630136987
실제값 : [0.001]

[628]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004512328767123287
실제값 : [0.004]

[652]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004506849315068493
실제값 : [0.004]

[664]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004506849315068493
실제값 : [0.004]

[745]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041095
실제값 : [0.006]

[785]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041096
실제값 : [0.006]

[837]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041096
실제값 : [0.006]

[962]번째 열들의 평균은 
마지막 열 직전까지의 값 0.006504109589041096
실제값 : [0.006]

[988]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004498630136986301
실제값 : [0.005]

[1206]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0034958904109589043
실제값 : [0.004]



해당 위치에서 mean 값이랑 값이 맞지 않습니다.

In [9]:
# 마지막 열의 값이 mena일까? 
test_lst= np.round(np.mean(view_rate.iloc[:-1,:].values,axis=0),3)==view_rate.iloc[-1,:].values
print(test_lst.all())
for i in np.argwhere(test_lst ==False):
    print("{}번째 열들의 평균은 ".format(i))
    print("마지막 열 직전까지의 값 {}".format(np.mean(view_rate.iloc[:-1,i].values)))
    print("실제값 : {}\n".format(view_rate.iloc[-1,i].values[0]))

False
[50]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0044777777777777784
실제값 : 0.005

[57]번째 열들의 평균은 
마지막 열 직전까지의 값 0.003486111111111111
실제값 : 0.004

[110]번째 열들의 평균은 
마지막 열 직전까지의 값 0.0024874999999999997
실제값 : 0.003

[184]번째 열들의 평균은 
마지막 열 직전까지의 값 0.003491666666666667
실제값 : 0.004

[204]번째 열들의 평균은 
마지막 열 직전까지의 값 0.00851875
실제값 : 0.008

[230]번째 열들의 평균은 
마지막 열 직전까지의 값 0.009490972222222224
실제값 : 0.01

[263]번째 열들의 평균은 
마지막 열 직전까지의 값 0.005501388888888889
실제값 : 0.005

[310]번째 열들의 평균은 
마지막 열 직전까지의 값 0.004525694444444444
실제값 : 0.004

[353]번째 열들의 평균은 
마지막 열 직전까지의 값 0.014594444444444441
실제값 : 0.014



---
## 의문점 쓰는 공간 

1. view_rate의 마지막 행과 열은 각각 무엇을 의미할까요???
> 마지막 열 : 93, 120, 628, 652, 745, 785, 837, 962, 988, 1206 위치에 있는 것들을 제외하고는 평균의 소수점 넷째자리에서 반올림한 값입니다. 
> <br>위의 예외들은 올린것도 있고...버린것도있고... 뒤죽박죽 
> <br>마지막 행 : 위와 마찬가지 ;;; 
